### PR TITLE
Riedl

### PR DESCRIPTION
**Pull request by jriedl**

### CODE REVIEW

File: cpp/pipelines/add_subscription_bundle_entries.cc [hunk 1]
```diff
@@ -37,18 +37,22 @@ using BundleToPPNsMap = std::map<std::string, std::set<std::string>>;
 
 
 [[noreturn]] void Usage() {
- std::cerr << "Usage: " << ::progname << "marc_input marc_output\n"
+ std::cerr << "Usage: " << ::progname << " marc_input marc_output\n"
```
Comment (ruschein) [inline cpp/pipelines/add_subscription_bundle_entries.cc:40]: Please use the Usage function from util.h.

File: cpp/pipelines/add_subscription_bundle_entries.cc [hunk 2]
```diff
@@ -37,18 +37,22 @@ using BundleToPPNsMap = std::map<std::string, std::set<std::string>>;
...
-MARC::Record GenerateBundleRecord(const std::string &record_id, const std::string &bundle_name, const std::vector<std::string> &instances) {
+MARC::Record GenerateBundleRecord(const std::string &record_id, const std::string &bundle_name, const std::vector<std::string> &instances,
+                                  const std::string &description)
+{
     const std::string today(TimeUtil::GetCurrentDateAndTime("%y%m%d"));
     // exclude from Ixtheo e.g. because it's a pure Relbib list
     const bool exclude_ixtheo(std::find(instances.begin(), instances.end(), "ixtheo") == instances.end());
     const bool include_relbib(std::find(instances.begin(), instances.end(), "relbib") != instances.end());
-    const bool include_bibstudies(std::find(instances.begin(), instances.end(), "bistudies") != instances.end() ? true : false);
+    const bool include_bibstudies(std::find(instances.begin(), instances.end(), "bibstudies") != instances.end() ? true : false);
```
Comment (ruschein) [inline cpp/pipelines/add_subscription_bundle_entries.cc:53]: Please get rid off the part starting with the question mark.  Also on the next statement.

File: cpp/pipelines/add_subscription_bundle_entries.cc [hunk 3]
```diff
@@ -119,7 +111,8 @@ void ProcessRecords(MARC::Reader * const marc_reader, MARC::Writer * const marc_
 
 int Main(int argc, char **argv) {
     if (argc < 3)
```
Comment (ruschein) [inline cpp/pipelines/add_subscription_bundle_entries.cc:113]: Shouldn't that be != 3 according to the usage mesage?

File: cpp/pipelines/add_subscription_bundle_entries.cc [hunk 4]
```diff
@@ -58,12 +54,17 @@ MARC::Record GenerateBundleRecord(const std::string &record_id, const std::strin
                                 { 'b', "1" /* series has not been completed */ } });
     record.insertField("935", { { 'c', "subskriptionspaket" } });
 
+    if (not description.empty())
+        record.insertField("500", { { 'a', description } });
```
Comment (ruschein) [inline cpp/pipelines/add_subscription_bundle_entries.cc:58]: There is a new insertField function that makes this easier.

File: cpp/lib/src/Downloader.cc [hunk 5]
```diff
@@ -711,6 +711,21 @@ void Downloader::setUserAgent(const std::string &user_agent) {
...
 
 
+bool Downloader::getHttpRedirectedUrl(std::string * const redirected_url) {
+    if (params_.follow_redirects_)
+        throw std::runtime_error(__FUNCTION__ + std::string(" Cannot determine redirect url if automatic redirection is enabled"));
+    //We only address 3XX HTTP codes
+    if (getResponseCode() / 100 != 3)
+        return false;
+    char *location(0);
+    curl_error_code_ = ::curl_easy_getinfo(getEasyHandle(), CURLINFO_REDIRECT_URL, &location);
```
Comment (ruschein) [inline cpp/lib/src/Downloader.cc:721]: Why don't you pass in redirected_url instead of &location?

File: cpp/lib/src/Downloader.cc [hunk 6]
```diff
@@ -711,6 +711,21 @@ void Downloader::setUserAgent(const std::string &user_agent) {
...
+    *redirected_url = location;
+    if (curl_error_code_ != CURLE_OK)
+       throw std::runtime_error(__FUNCTION__ + std::string(" Could not determine redirect url"));
```
Comment (ruschein) [inline cpp/lib/src/Downloader.cc:724]: "url" is an acronym and should be capitalized.

File: cpp/lib/src/SmartDownloader.cc [hunk 7]
```diff
@@ -359,3 +366,13 @@ bool SmartDownload(const std::string &url, const TimeLimit &time_limit, std::str
...
     return false;
 }
+
+
+bool SmartDownloadResolveFirstRedirectHop(const std::string &url, const TimeLimit &time_limit, std::string * const document,
+                   std::string * const http_header_charset, std::string * const error_message,
+                   const bool trace)
+{
+    std::string redirected_url(url);
+    GetRedirectedUrl(url, time_limit, &redirected_url);
```
Comment (ruschein) [inline cpp/lib/src/SmartDownloader.cc:376]: why are you not testing the return code here?

File: cpp/pipelines/flag_records_as_available_in_tuebingen.cc [hunk 8]
```diff
@@ -59,18 +54,26 @@ void ProcessSuperiorRecord(const MARC::Record &record, RegexMatcher * const tue_
...
+            }
         }
     }
 }
 
 
-void LoadDE21PPNs(MARC::Reader * const marc_reader, RegexMatcher * const tue_sigil_matcher, std::unordered_set<std::string> * const de21_superior_ppns,
-                  unsigned * const extracted_count)
+void LoadDE21AndTADPPNs(MARC::Reader * const marc_reader,
+                  RegexMatcher * const tue_sigil_matcher, std::unordered_set<std::string> * const de21_superior_ppns,
```
Comment (ruschein) [inline cpp/pipelines/flag_records_as_available_in_tuebingen.cc:67]: Indentation looks wrong.

File: cpp/pipelines/flag_records_as_available_in_tuebingen.cc [hunk 9]
```diff
@@ -59,18 +54,26 @@ void ProcessSuperiorRecord(const MARC::Record &record, RegexMatcher * const tue_
...
+                  unsigned * const extracted_count, unsigned * const extracted_tad_count)
 {
     while (const MARC::Record record = marc_reader->read())
-         ProcessSuperiorRecord(record, tue_sigil_matcher, de21_superior_ppns, extracted_count);
+         ProcessSuperiorRecord(record, tue_sigil_matcher, de21_superior_ppns, tad_sigil_matcher, tad_superior_ppns,
+                               extracted_count, extracted_tad_count) ;
 
-    LOG_DEBUG("Finished extracting " + std::to_string(*extracted_count) + " superior records");
+    LOG_DEBUG("Finished extracting " + std::to_string(*extracted_count) + " superior records and " +
+               std::to_string(*extracted_tad_count) + " TAD supeprior records");
```
Comment (ruschein) [inline cpp/pipelines/flag_records_as_available_in_tuebingen.cc:76]: typo

File: cpp/pipelines/flag_records_as_available_in_tuebingen.cc [hunk 10]
```diff
@@ -96,9 +99,9 @@ void CollectSuperiorPPNs(const MARC::Record &record, std::unordered_set<std::str
 }
 
 
-void FlagRecordAsInTuebingenAvailable(MARC::Record * const record, unsigned * const modified_count) {
-    record->insertField("ITA", { { 'a', "1" } });
-     ++*modified_count;
+void FlagRecordAsInTuebingenAvailable(MARC::Record * const record, const bool tad_available, unsigned * const modified_count) {
+    tad_available ? record->insertField("ITA", {  { 'a', "1" }, { 't', "1" } }) : record->insertField("ITA", { { 'a', "1" } });
```
Comment (ruschein) [inline cpp/pipelines/flag_records_as_available_in_tuebingen.cc:103]: We could use the simplified insertField for the 2nd insertField here.

File: cpp/pipelines/flag_records_as_available_in_tuebingen.cc [hunk 11]
```diff
@@ -118,10 +121,13 @@ bool AlreadyHasLOK852DE21(const MARC::Record &record, RegexMatcher * const tue_s
 }
 
 
-void ProcessRecord(MARC::Record * const record, MARC::Writer * const marc_writer, RegexMatcher * const tue_sigil_matcher,
-                   std::unordered_set<std::string> * const de21_superior_ppns, unsigned * const modified_count) {
+void ProcessRecord(MARC::Record * const record, MARC::Writer * const marc_writer,
+                   RegexMatcher * const tue_sigil_matcher, std::unordered_set<std::string> * const de21_superior_ppns,
+                   RegexMatcher * const tad_sigil_matcher, std::unordered_set<std::string> * const tad_superior_ppns,
+                   unsigned * const modified_count) {
```
Comment (ruschein) [inline cpp/pipelines/flag_records_as_available_in_tuebingen.cc:127]: The opening brace belongs on the following line.

File: cpp/pipelines/flag_records_as_available_in_tuebingen.cc [hunk 12]
```diff
@@ -134,24 +140,39 @@ void ProcessRecord(MARC::Record * const record, MARC::Writer * const marc_writer
...
     marc_writer->write(*record);
 }
 
 
 void AugmentRecords(MARC::Reader * const marc_reader, MARC::Writer * const marc_writer,
                     RegexMatcher * const tue_sigil_matcher, std::unordered_set<std::string> * const de21_superior_ppn,
-                    unsigned * const extracted_count, unsigned * const modified_count) {
+                    RegexMatcher * const tad_sigil_matcher, std::unordered_set<std::string> * const tad_superior_ppn,
+                    unsigned * const extracted_count, unsigned * const extracted_tad_count,
+                    unsigned * const modified_count) {
```
Comment (ruschein) [inline cpp/pipelines/flag_records_as_available_in_tuebingen.cc:169]: brace!

File: cpp/lib/src/Downloader.cc (hunk 5 above)
Comment (jriedl) [reply cpp/lib/src/Downloader.cc:721]: Apparently curl_easy_getinfo allocates and manages memory internally

Comment (ruschein) [reply cpp/lib/src/Downloader.cc:721]: Accepted, but, don't we have to release this memory?

Comment (jriedl) [reply cpp/lib/src/Downloader.cc:721]: According to https://curl.haxx.se/libcurl/c/curl_easy_getinfo.html obviously not.

Comment (jriedl) [reply cpp/lib/src/Downloader.cc:721]: i.e. is seemingly done by curl_easy_cleanup in the destructor.

File: cpp/lib/src/Downloader.cc [hunk 13]
```diff
@@ -711,6 +711,21 @@ void Downloader::setUserAgent(const std::string &user_agent) {
 }
 
 
+bool Downloader::getHttpRedirectedUrl(std::string * const redirected_url) {
+    if (params_.follow_redirects_)
+        throw std::runtime_error(__FUNCTION__ + std::string(" Cannot determine redirect url if automatic redirection is enabled"));
+    //We only address 3XX HTTP codes
+    if (getResponseCode() / 100 != 3)
+        return false;
+    char *location(0);
```
[the remark this answers is not in the document]
Comment (ruschein) [reply cpp/lib/src/Downloader.cc:720]: We also need a comment here as to why we even need this variable.

File: cpp/lib/src/BSZUtil.cc [hunk 14]
```diff
@@ -192,11 +192,14 @@ void ExtractArchiveMembers(const std::string &archive_name, std::vector<std::str
 
 
 void ExtractYearVolumeIssue(const MARC::Record &record, std::string * const year, std::string * const volume, std::string * const issue) {
+    const auto field_008(record.findTag("008"));
+    if (field_008 != record.end())
+         *year = field_008->getContents().substr(7,4);
```
Comment (ruschein) [inline cpp/lib/src/BSZUtil.cc:197]: Missing comma

**Pull request merged**